### ISSUE TITLE
Missing /Contents/Home/bundle/libjimage.dylib in AdoptOpenJDK 14 when running Eclipse from command line in macOS

### ISSUE THREAD

**Issue by deors** [AdoptOpenJDK/openjdk-support#123]
**Platform:** / **Architecture:**
macOS Catalina 10.15.5
AdoptOpenJDK 14.0.1+7 (acquired through Homebrew)
Eclipse 2020-03
JDK Mission Control 8 (from AdoptOpenJDK)

When executing eclipse executable or jmc executable from command line, JVM fails and exists, showing this error:

Error occurred during initialization of VM
Unable to load jimage library: /Library/Java/JavaVirtualMachines/adoptopenjdk-14.jdk/Contents/Home/bundle/libjimage.dylib

This error does not happen when both applications are launched from Launchpad. I have confirmed with JDK Mission Control that indeed the same AdoptOpenJDK installation is being used.

I have also test with other JDK 14 distributions, and the error is not happening.

Finally, I have verified that libjimage.dylib is indeed there, but in /lib folder as in the other distros, but not in /bundle folder.

I'm happy to provide with more information or do a screen sharing session if needed.

**Comment by aahlenst** [comment]
What other distributions have you tested with? It's a bit weird, because the others I have lying around have `libjimage.dylib` all in the same place as we do and do not have one in the `bundle` folder:

```
$ find /Library/Java/JavaVirtualMachines/adoptopenjdk-14.jdk -name "libjimage*" 
/Library/Java/JavaVirtualMachines/adoptopenjdk-14.jdk/Contents/Home/lib/libjimage.dylib
$ find zulu14.28.21-ca-jdk14.0.1-macosx_x64 -name "libjimage*" 
zulu14.28.21-ca-jdk14.0.1-macosx_x64/zulu-14.jdk/Contents/Home/lib/libjimage.dylib
$ find jdk-14.0.1.jdk -name "libjimage*"         
jdk-14.0.1.jdk/Contents/Home/lib/libjimage.dylib
jdk-14.0.1.jdk/lib/libjimage.dylib
```

(The last one is BellSoft Liberica)

**Comment by tresf** [comment]
> AdoptOpenJDK 14.0.1+7 (acquired through Homebrew)

Off topic, but how?  I can't find it and the PR [hasn't been merged yet](https://github.com/Homebrew/homebrew-core/pull/51883).

**Comment by karianna** [comment]
> > AdoptOpenJDK 14.0.1+7 (acquired through Homebrew)
> 
> Slightly off-topic, but how? I can't find it and the PR [hasn't been merged yet](https://github.com/Homebrew/homebrew-core/pull/51883).

You'll need to tap our cask in our homebrew repo now.

**Comment by tresf** [comment]
It's working for me on Big Sur:



```
brew cask install adoptopenjdk14
brew tap AdoptOpenJDK/openjdk
# will ask for password

# remove quarantine flag
sudo xattr -r -d com.apple.quarantine org.openjdk.jmc-8.0.0-SNAPSHOT-macosx.cocoa.x86_64/JDK\ Mission\ Control.app/

# start jmc
org.openjdk.jmc-8.0.0-SNAPSHOT-macosx.cocoa.x86_64/JDK\ Mission\ Control.app/Contents/MacOS/jmc              
OpenJDK 64-Bit Server VM warning: Option FlightRecorder was deprecated in version 13.0 and will likely be removed in a future release.
Started recording 1.

... 
```

**Comment by karianna** [comment]
I also can't reproduce this :-( on 10.15.5

**Comment by tresf** [comment]
> I also can't reproduce this :-( on 10.15.5

Just a hunch, but I have a sneaking suspicion that the OP has found a way to register the `jmc` command which isn't provided in the original bug report:

```diff
  user@mac % jmc
- Unable to locate an executable at "/Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home/bin/jmc": No such file or directory
```

**Comment by deors** [comment]
I will try it again with a clear installation later tonight and let you
know how it's going. BR, J.

**Comment by aahlenst** [comment]
@tresf Homebrew is a bit a mess. There is `adoptopenjdk` in homebrew-casks that is on 14.0.1 and notarized. That is maintained by us and always points to the latest release and auto-updates to 15 in September. To install, you need `brew cask install adoptopenjdk`. For fixed versions, you need our tap as Martijn has explained.

**Comment by tresf** [comment]
@aahlenst 

> @tresf Homebrew is a bit a mess. There is adoptopenjdk in homebrew-casks that is on 14.0.1 and notarized. That is maintained by us and always points to the latest release and auto-updates to 15 in September. To install, you need brew cask install adoptopenjdk. For fixed versions, you need our tap as Martijn has explained.

Thanks, I realized after @karianna's response that I was using `openjdk14` not `adoptopenjdk14`.  In regards to it. being a `mess`, I'd argue that there are advantages of `openjdk` as it doesn't require `sudo` for installation, but that's most certainly off-topic so I won't carry on :).  The cask worked as expected. 👍   Additionally, the cask offers a path to use adopt in CI that I hadn't realized was possibly prior, so I appreciate the explanation.

**Comment by deors** [comment]
Hi, I've been testing this again. AdoptOpenJDK is installed from adoptopenjdk cask (latest) and installs into /Library/Java/JavaVirtualMachines/adoptopenjdk-14.jdk/. After installation, the libjimage.dylib is not present in the expected directory Contents/Home/bundle/libjimage.dylib.
However, I cannot understand why launching JMC or Eclipse from Launchpad works fine, and do not complain about that missing file.
Regarding this:
> Just a hunch, but I have a sneaking suspicion that the OP has found a way to register the `jmc` command which isn't provided in the original bug report:
I have not registered that command, I'm launching jmc from the directory where it is installed: /Applications/JDK\ Mission\ Control\ 8.app/Contents/MacOS/jmc

Any other clues at what might probably be happening? Happy to try and test as per your suggestions.

BR, J.

**Comment by aahlenst** [comment]
@reinhapa @thegreystone Looks like JMC is looking in the wrong place. Could one of you have a look?

**Comment by thegreystone** [comment]
There was a bug in JDK 14.0.0 which made it impossible to start Eclipse (and thus any Eclipse RCP based applications) on Mac. That issue was fixed in 14.0.1 (and backported to various versions where the same change had been made):
https://bugs.openjdk.java.net/browse/JDK-8238225

**Comment by deors** [comment]
Hi,
I'm on 14.0.1+7 already.
Actually as explained, Eclipse or JMC runs well when launched from Launchpad. The error only appears when launched from CLI.
And also important to note, other OpenJDK distros that I've checked, has that library in the expected directory. Could this be something that is not generated well within the dmg package?
BR, J.

**Comment by aahlenst** [comment]
@deors Which distributions did you check? Because, as I wrote earlier:

> It's a bit weird, because the others I have lying around have `libjimage.dylib` all in the same place as we do and do not have one in the `bundle` folder:
> 
> ```
> $ find /Library/Java/JavaVirtualMachines/adoptopenjdk-14.jdk -name "libjimage*" 
> /Library/Java/JavaVirtualMachines/adoptopenjdk-14.jdk/Contents/Home/lib/libjimage.dylib
> $ find zulu14.28.21-ca-jdk14.0.1-macosx_x64 -name "libjimage*" 
> zulu14.28.21-ca-jdk14.0.1-macosx_x64/zulu-14.jdk/Contents/Home/lib/libjimage.dylib
> $ find jdk-14.0.1.jdk -name "libjimage*"         
> jdk-14.0.1.jdk/Contents/Home/lib/libjimage.dylib
> jdk-14.0.1.jdk/lib/libjimage.dylib
> ```
> 
> (The last one is BellSoft Liberica)

**Comment by deors** [comment]
Hi, you're correct, I did a bad job explaining. They do not have the library in that place, but they do not complain about it, and both Eclipse IDE and JMC launch well both from terminal and Launchpad.
I've tested with Zulu and Java.net OpenJDK distros.

**Comment by deors** [comment]
So, looking all together. Why is the Eclipse and JMC executable complain about the library in that specific folder, only in AdoptOpenJDK distro? What is different, and what is causing the exec to look for the library in the wrong folder? And why only when launched in the terminal?

**Comment by deors** [comment]
If there is anything I can do on my end to debug/profile this, don't hesitate to let me know.

**Comment by aahlenst** [comment]
@deors Let's try to isolate the problem.

What I did as preparation:

```
$ wget https://ci.adoptopenjdk.net/view/JMC/job/jmc-build/job/7.x.x/lastSuccessfulBuild/artifact/target/products/org.openjdk.jmc-7.1.1-macosx.cocoa.x86_64.tar.gz
$ tar xvfz org.openjdk.jmc-7.1.1-macosx.cocoa.x86_64.tar.gz
$ export JAVA_HOME=$(/usr/libexec/java_home -v 14)
$ java -version
openjdk version "14.0.1" 2020-04-14
OpenJDK Runtime Environment AdoptOpenJDK (build 14.0.1+7)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 14.0.1+7, mixed mode, sharing)
```

What I tested afterwards:

* Double clicking JDK Mission Control.app: ✅
* `open JDK\ Mission\ Control.app`: ✅
* `./JDK\ Mission\ Control.app/Contents/MacOS/jmc`: ✅

About my system:

```
$ system_profiler SPSoftwareDataType
Software:

    System Software Overview:

      System Version: macOS 10.15.5 (19F101)
      Kernel Version: Darwin 19.5.0
      Boot Volume: Macintosh HD
      Boot Mode: Normal
      Computer Name: redacted
      User Name: redacted
      Secure Virtual Memory: Enabled
      System Integrity Protection: Enabled
      Time since boot: 13 minutes
```

So, I cannot reproduce your problem, but that's not that important at the moment. What exactly did you do? Do the things I did work for you, too?

**Comment by deors** [comment]
Hi Andreas.
The main difference I see is that I'm working with JMC 8.0.0 latest
snapshot, also from AdoptOpenJDK project.
Let me try later with 7.1.1 and will come back to you with the findings.
Thanks a lot,
J.

**Comment by aahlenst** [comment]
```
$ wget https://ci.adoptopenjdk.net/view/JMC/job/jmc-build/job/master/lastSuccessfulBuild/artifact/target/products/org.openjdk.jmc-8.0.0-SNAPSHOT-macosx.cocoa.x86_64.tar.gz
$ tar xvfz org.openjdk.jmc-8.0.0-SNAPSHOT-macosx.cocoa.x86_64.tar.gz
$ export JAVA_HOME=$(/usr/libexec/java_home -v 14)
$ java -version
openjdk version "14.0.1" 2020-04-14
OpenJDK Runtime Environment AdoptOpenJDK (build 14.0.1+7)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 14.0.1+7, mixed mode, sharing)
$ JDK\ Mission\ Control.app/Contents/MacOS/jmc -version
OpenJDK 64-Bit Server VM warning: Option FlightRecorder was deprecated in version 13.0 and will likely be removed in a future release.
Started recording 1.

Use jcmd 3963 JFR.dump name=JMC_Default filename=FILEPATH to copy recording data to file.
========================================
JDK Mission Control 8.0.0
  java.vm.version = 14.0.1+7
  jmc.qualifier = 202007040205
  jmc.buildid = jenkins
  jmc.changeid = bfdc3eb0756f92dc5f5c81b50c83b9f15364a19d
  jmc.date = 2020-07-04
========================================
```

Results:

* Double clicking JDK Mission Control.app: ✅
* `open JDK\ Mission\ Control.app`: ✅
* `./JDK\ Mission\ Control.app/Contents/MacOS/jmc`: ✅

**Comment by aahlenst** [comment]
@deors Have you been able to have a look what we're doing differently?

**Comment by deors** [comment]
hi @aahlenst,
I've tried again, following exactly the same commands, for both 7.1.1 and 8.0.0. In both cases, I get the same error:
```
Error occurred during initialization of VM
Unable to load jimage library: /Library/Java/JavaVirtualMachines/adoptopenjdk-14.jdk/Contents/Home/bundle/libjimage.dylib
```
To be precise - Results:
- Double clicking JDK Mission Control.app: nothing happens (silent error?)
- ```open JDK\ Mission\ Control.app```: nothing happens (silent error?)
- ```./JDK\ Mission\ Control.app/Contents/MacOS/jmc```: the reported error is shown in terminal
I'm running out of ideas, so any other pointers and what to look will be greatly appreciated.
Thanks for all the help so far.
BR, J.

**Comment by tresf** [comment]
@deors,

What's the value of `$PATH` on your machine?  Mine's `/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin`

**Comment by deors** [comment]
hi @aahlenst and rest of team,
I think that I've isolated the root cause. I've checked again everything, and found an important detail in the environment: I am using jenv to easily switch between different Java runtimes.
Although at first sight should do no harm - just run the script and get some symlinks updated to point to the desired JDK, and although it works in every other scenario I had faced so far, this one specifically fails.
As per the root cause, jenv puts a java exec in the path which is not really a symlink, but a custom script. The script itself is a bit strange as it ends executing java inside its own binary. Not sure why that is needed (a question to ask to jenv team), but reality is that if I overwrite path to put $JAVA_HOME/bin upfront, JMC, Eclipse works as expected.

Path with error was: /Users/xxx/.jenv/shims:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
Forced path: /Library/Java/JavaVirtualMachines/adoptopenjdk-14.jdk/Contents/Home/bin:/Users/xxx/.jenv/shims:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin

```
#!/usr/bin/env bash
set -e
[ -n "$JENV_DEBUG" ] && set -x

program="${0##*/}"
if [ "$program" = "java" ]; then
  for arg; do
    case "$arg" in
    -e* | -- ) break ;;
    */* )
      if [ -f "$arg" ]; then
        export JENV_DIR="${arg%/*}"
        break
      fi
      ;;
    esac
  done
fi

export JENV_ROOT="/Users/jorgehidalgo/.jenv"
exec "/usr/local/Cellar/jenv/0.5.3/libexec/libexec/jenv" exec "$program" "$@"
```

So, in a nutshell, it seems that the root cause for the issue is within jenv. I will try to investigate a bit more and share the findings with jenv team.
Thank you very much for all the help, and sincere apologies for not spotting the difference in the environment before.
BR, J.

**Comment by deors** [comment]
hi @tresf,
Thanks for the pointer. Our comments crossed (I was writing while your comment came to live) but as explained above the issue lies with path, specifically path poluted by jenv with a custom script.
BR, J.
